spring-data-relational/src/main/java/org/springframework/data/relational/core/sql/CaseExpression.java

```diff
@@ -0,0 +1,94 @@
+package org.springframework.data.relational.core.sql;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.util.stream.Collectors.joining;
+
+/**
+ * Case with one or more conditions expression.
+ * <p>
+ * Results in a rendered condition:
+ * <pre>
+ *   CASE
+ *     WHEN condition1 THEN result1
+ *     WHEN condition2 THEN result2
+ *     ELSE result
+ *   END
+ * </pre>
+ * </p>
+ *
+ * @author Sven Rienstra
+ * @since 3.4
+ */
+public class CaseExpression extends AbstractSegment implements Expression {
+    private final List<When> whenList = new ArrayList<>();
+    private final Literal other;
+
+    private CaseExpression(List<When> whenList, Literal other) {
+
+        super(children(whenList, other));
+
+        this.whenList.addAll(whenList);
+        this.other = other;
+    }
+
+    /**
+     * Create CASE {@link Expression} with initial {@link When} condition.
+     * @param condition initial {@link When} condition
+     * @return the {@link CaseExpression}
+     */
+    public static CaseExpression create(When condition) {
+        return new CaseExpression(List.of(condition), null);
+    }
+
+    /**
+     * Add additional {@link When} condition
+     * @param condition the {@link When} condition
+     * @return the {@link CaseExpression}
+     */
+    public CaseExpression when(When condition) {
+        this.whenList.add(condition);
+        return new CaseExpression(whenList, other);
```

Comment:
This looks like it uses `CaseExpression` as an immutable, but it actually modifies the `whenList`. We should avoid that modification.